src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameService.cs

```diff
@@ -84,6 +91,41 @@ internal class InlineRenameService : IInlineRenameService
                 _asyncListener);
 
             return new InlineRenameSessionInfo(ActiveSession);
+
+            static bool IsReadOnlyOrCannotNavigateToSpan(IInlineRenameInfo renameInfo, Document document, CancellationToken cancellationToken)
+            {
+                // This is unpleasant, but we do everything synchronously.  That's because we end up
+                // needing to make calls on the UI thread to determine if the locations of the symbol
+                // are in readonly buffer sections or not.  If we go pure async we have the following
+                // problem:
+                //   1) if we call ConfigureAwait(false), then we might call into the text buffer on 
+                //      the wrong thread.
+                //   2) if we try to call those pieces of code on the UI thread, then we will deadlock
+                //      as our caller often is doing a 'Wait' on us, and our UI calling code won't run.
+                if (renameInfo is SymbolInlineRenameInfo symbolInlineRenameInfo)
+                {
+                    var workspace = document.Project.Solution.Workspace;
+                    var navigationService = workspace.Services.GetService<IDocumentNavigationService>();
+                    foreach (var documentSpan in symbolInlineRenameInfo.DocumentSpans)
+                    {
+                        var sourceText = documentSpan.Document.GetTextAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+                        var textSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
+
+                        if (textSnapshot != null)
+                        {
+                            var buffer = textSnapshot.TextBuffer;
+                            var originalSpan = documentSpan.SourceSpan.ToSnapshotSpan(textSnapshot).TranslateTo(buffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+
+                            if (buffer.IsReadOnly(originalSpan) || !navigationService.CanNavigateToSpan(workspace, document.Id, documentSpan.SourceSpan))
+                            {
+                                return true;
+                            }
```

Comment:
nit: we could use this opportunity to produce distinct failure messages for these two cases.  Specifically "You_cannot_rename_this_element"is terrible.  It would be better as "You_cannot_rename_this_element_because_it_is_contained_in_a_read_only_file" or "You_cannot_rename_this_element_because_it_is_in_a_location_that_cannot_be_navigated_to".  (can be done in the future).